docs/build/voting-dapp-tutorial.md

```diff
@@ -0,0 +1,1245 @@
+      (enforce (= exists true) "Candidate doesn't exist"))
+
+    (with-capability (ACCOUNT-OWNER account)
+      (vote-protected account cid))
+
+    (format "Voted for candidate {}!" [cid])
+  )
+```
+
+Quite a lot happening here, so let's go line by line:
+* to prevent double voting, we have to check if the user already voted. Below is the implementation of the `user-voted` function which does that.
+```clojure
+(defun user-voted:bool (account:string)
+  (with-default-read votes account
+    { "cid": "" }
+    { "cid":= cid }
+    (> (length cid) 0))
+)
+```
+* to prevent voting for a candidate that doesn't exist we implemented `candidate-exists` function and we `enforce` it returns `true` for our parameters
+
+```clojure
+(defun candidate-exists:bool (cid:string)
+  (with-default-read candidates cid
+    { "name": "" }
+    { "name" := name }
+    (> (length name) 0))
+)
+```
+
+:::note
+Notice how we used a similar approach in both `user-voted` and `candidate-exists` functions to execute a single row read using `with-default-read`. Below is an alternative, naive a way to implement the same thing:
+
+```clojure
+(contains account (keys votes))
+```
+The problem with this approach is that `keys` function queries all rows in the table and it's a very expensive operation. Do not use it.
+:::
+
+* we are trying to acquire the `ACCOUNT-OWNER` capability which confirms that who submitted the transaction is the owner of the KDA account. Below is the implementation of the capability:
+```clojure
+; import coin.details function
+(use coin [ details ])
+
+(defcap ACCOUNT-OWNER (account:string)
+  "Make sure the requester owns the KDA account"
+  (enforce-guard (at 'guard (coin.details account)))
+)
+```
+* while the `ACCOUNT-OWNER` capability is in scope we are calling `vote-protected` which is the function that updates the database
+```clojure
+(defun vote-protected (account:string candidateId:string)
+  (require-capability (ACCOUNT-OWNER account))
+
+  (with-read candidates candidateId { "votes" := votesCount }
+    (update candidates candidateId { "votes": (+ votesCount 1) })
+    (insert votes account { "cid": candidateId })
+    (emit-event (VOTED candidateId))
+  )
+)
+```
+First it checks that we have the capability, then proceed to update the votes of a candidate and record the vote in the votes table. At the end we emit the `VOTED` event which is useful for our frontend to update itself in real-time by listening to this event.
+
+:::tip
+Spend a bit of time understanding the logic in the `vote` function before moving forward.
+:::
+
+We're missing a way to read the current number of votes so here it is:
+
+```clojure
+(defun getVotes:integer (cid:string)
+  "Get the votes count by cid"
+  (at 'votes (read candidates cid ['votes]))
+)
+```
+
+And we also need to add some candidates in our table:
+
+```clojure
+(defun init ()
+  "Initialize the rows in candidates table"
+  (insert candidates "1" { "name": "Candidate A", "votes": 0 })
```

Comment:
Can we add an `insert-candidate` function that can only be performed by admin? Perhaps it takes a list so that it can be done in bulk? Would be a good way to demonstrate govt / guarding code